cirq-core/cirq/sim/classical_simulator.py

```diff
@@ -12,96 +12,211 @@
-        param_resolver: Parameters to run with the program.
-        repetitions: Number of times to repeat the run. It is expected that
-            this is validated greater than zero before calling this method.
-
-    Returns:
-        A dictionary mapping measurement keys to measurement results.
-
-    Raises:
-        ValueError: If
-            - one of the gates is not an X, CNOT, SWAP, TOFFOLI or a measurement.
-            - A measurement key is used for measurements on different numbers of qubits.
-    """
-
-    def _run(
-        self, circuit: AbstractCircuit, param_resolver: ParamResolver, repetitions: int
-    ) -> Dict[str, np.ndarray]:
-        results_dict: Dict[str, np.ndarray] = {}
-        values_dict: Dict[Qid, int] = defaultdict(int)
-        param_resolver = param_resolver or ParamResolver({})
-        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
-
-        for moment in resolved_circuit:
-            for op in moment:
-                if _is_identity(op):
-                    continue
-                if op.gate == ops.X:
-                    (q,) = op.qubits
-                    values_dict[q] ^= 1
-                elif op.gate == ops.CNOT:
-                    c, q = op.qubits
-                    values_dict[q] ^= values_dict[c]
-                elif op.gate == ops.SWAP:
-                    a, b = op.qubits
-                    values_dict[a], values_dict[b] = values_dict[b], values_dict[a]
-                elif op.gate == ops.TOFFOLI:
-                    c1, c2, q = op.qubits
-                    values_dict[q] ^= values_dict[c1] & values_dict[c2]
-                elif protocols.is_measurement(op):
-                    measurement_values = np.array(
-                        [[[values_dict[q] for q in op.qubits]]] * repetitions, dtype=np.uint8
-                    )
-                    key = op.gate.key  # type: ignore
-                    if key in results_dict:
-                        if op._num_qubits_() != results_dict[key].shape[-1]:
-                            raise ValueError(
-                                f'Measurement shape {len(measurement_values)} does not match '
-                                f'{results_dict[key].shape[-1]} in {key}.'
-                            )
-                        results_dict[key] = np.concatenate(
-                            (results_dict[key], measurement_values), axis=1
-                        )
-                    else:
-                        results_dict[key] = measurement_values
-                else:
-                    raise ValueError(
-                        f'{op} is not one of cirq.X, cirq.CNOT, cirq.SWAP, '
-                        'cirq.CCNOT, or a measurement'
-                    )
-
-        return results_dict
+class ClassicalBasisState(qis.QuantumStateRepresentation):
+    """Represents a classical basis state for efficient state evolution."""
+
+    def __init__(self, initial_state: List[int]):
```

Comment:
Hi @NoureldinYosri, I'm hitting a type issue when indexing `self._state.basis` with type np.ndarray. 

Originally I assumed the basis would always be a `Sequence[int]` and would index like this:

```python
def _act_on_fallback_(self, action, qubits: Sequence['cirq.Qid'], allow_decompose: bool = True):
        gate = action.gate if isinstance(action, ops.Operation) else action
        mapped_qubits = [self.qubit_map[i] for i in qubits]
        if gate == ops.CNOT:
            c, q = mapped_qubits
            self._state.basis[q] ^= self._state.basis[c]
        else:
            raise ValueError(f'{gate} is not one of X, CNOT, SWAP, CCNOT, or a measurement')
        return True
```

Since the basis could be a `nd.ndarray`, is it fair to assume that it's dimension could be >= 0? 
- If so how should I index `self._state.basis` using the `qid` for an N dimensional state with numpy?


At first I assumed that the qubits would be along the diagonal of the matrix: 
```python

self._state.basis[q][q][q]... ^= self._state.basis[c][c][c]...

```

I am not sure if this is a valid assumption for classical simulator?

Reply:
the statevector is 1D. you can add that as a validation step to the constructor if you want.
```py3
if len(state_vector.shape) != 1: raise ValueError(...)
```
 
btw what we are representing is not strictly the state vector which would have a dimension $2^\textit{num qubits}$ but a factorization of it which you call `_state.basis` and has dimension `num qubits`. we can do this here because there is no possible entanglement in classical simulation